structure_tests/drivers/docker_driver.go

```diff
@@ -0,0 +1,341 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package drivers
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"context"
+
+	"bufio"
+
+	"regexp"
+
+	"github.com/GoogleCloudPlatform/runtimes-common/structure_tests/types/unversioned"
+	"github.com/fsouza/go-dockerclient"
+)
+
+type DockerDriver struct {
+	cli           docker.Client
+	originalImage string
+	currentImage  string
+}
+
+func (d *DockerDriver) Info() string {
+	return fmt.Sprintf("DockerDriver:\nOriginalImage: %s\ncurrentImage: %s\ncli: %s\n",
+		d.originalImage, d.currentImage, d.cli)
+}
+
+func NewDockerDriver(image string) *DockerDriver {
+	newCli, err := docker.NewClientFromEnv()
+	if err != nil {
+		panic(err)
+	}
+	return &DockerDriver{
+		originalImage: image,
+		currentImage:  image,
+		cli:           *newCli,
+	}
+}
+
+func (d *DockerDriver) ProcessCommand(t *testing.T, envVars []unversioned.EnvVar, fullCommand []string,
+	shellMode bool, checkOutput bool) (string, string, int) {
+
+	if len(fullCommand) == 0 {
+		t.Logf("empty command provided: skipping...")
+		return "", "", -1
+	}
+	var env []string
+	for _, envVar := range envVars {
+		env = append(env, envVar.Key+"="+envVar.Value)
+	}
+	stdout, stderr, exitCode := d.exec(t, env, fullCommand)
+
+	if stdout != "" {
+		t.Logf("stdout: %s", stdout)
+	}
+	if stderr != "" {
+		t.Logf("stderr: %s", stderr)
+	}
+	//reset image for next test
+	d.currentImage = d.originalImage
+	return stdout, stderr, exitCode
+}
+
+func (d *DockerDriver) SetEnvVars(t *testing.T, vars []unversioned.EnvVar) []unversioned.EnvVar {
+	if len(vars) == 0 {
+		return nil
+	}
+
+	image, err := d.cli.InspectImage(d.currentImage)
+	if err != nil {
+		t.Errorf("Error when inspecting image: %s", err.Error())
+		return nil
+	}
+
+	// convert env to map for easier processing
+	imageEnv := make(map[string]string)
+	for _, varPair := range image.Config.Env {
+		pair := strings.Split(varPair, "=")
+		imageEnv[pair[0]] = pair[1]
+	}
+
+	before := regexp.MustCompile(".*\\$(.*?):")
+	after := regexp.MustCompile(".*:\\$(.*)")
+
+	env := []string{}
+	for _, envVar := range vars {
+		currentVar := ""
+		if match := before.FindStringSubmatch(envVar.Value); match != nil {
+			// first entry is the leftmost substring: second entry is the first group
+			currentVar = match[1]
+		}
+		if match := after.FindStringSubmatch(envVar.Value); match != nil {
+			currentVar = match[1]
+		}
+		if currentVar != "" {
+			if val, ok := imageEnv[currentVar]; ok {
+				env = append(env, envVar.Key+"="+strings.Replace(envVar.Value, "$"+currentVar, val, -1))
+			} else {
+				t.Errorf("Variable %s not found in image env! Check test config.", currentVar)
+			}
+		} else {
+			env = append(env, envVar.Key+"="+envVar.Value)
+		}
+	}
+	// since these are global envvars, just overwrite the original image
+	d.originalImage = d.runAndCommit(t, env, nil)
```

Comment:
If you reset current to original at the start, you can avoid overwriting original I think.

Reply:
I think this is necessary, because this is setting the global env vars for the entire test run. we want the original image to actually be a modified version of the original image with the global env vars set